### ISSUE TITLE
aarch64: VM panics for " ubuntu bash" 

### ISSUE THREAD

**Issue by chl337** [QuarkContainer/Quark#1161]
# Issue:

The VM panics while trying to run bash on an Ubuntu image.
The reported reason is: _exception 0x3C on EL1 - Software Breakpoint Instruction exception_.

# Current Build:
* #1133
* #1131
* #1106
* PAN unset in Memcpy

# Test
docker run --rm -it --runtime=quark_d ubuntu bash

---

I suspect a misalignment issue that leads to the wrong opcode, but I am open to suggestions.

**Comment by shrik3** [comment]
have you inspected the bash binary wrt. the faulting PC?

**Comment by chl337** [comment]
With some more debugging:
`
ESR_EL1: 0xF2000001
`
`
PC: 0x40003E599C -> NOTE: it will vary slightly depending on the current build
`
`
PC[opcode]: 0xD400020 => brk #0x1
`
The PC lies in the Kernel, so dumping the kernel points to
`
<_ZN61 ... qkernel..qlib..SysCallID$ ... more mangle ... *fmt..Debug*...>
`
The error is that the current _callId_ is not part of the declared _SysCallID_ definitions, while it is still less than the upper bound _UnkownSyscall_.

I will provide fixies.

**Comment by shrik3** [comment]
breakpoints should be ignored at this point. Apply this if you haven't

Btw. with the patch below the code proceeds to panic at 
```
[INFO] [1/43c0180000|0] (1/1)------get call id sys_io_setup arg0:ffffffffffffff9c
[ERROR] [1/43c0180000|0] NotImplementSyscall syscall SyscallArguments { arg0: ffffffffffffff9c, arg1: e99e0ce0520, arg2: 4, arg3: 200, arg4: 2ac47b5f6b58, arg5: ffffffffffffff9c }
[Print] [1/43c0180000|0] get panic : Some(Sycall not implement syscall is sys_io_setup)
```

```
diff --git a/qkernel/src/interrupt/aarch64/mod.rs b/qkernel/src/interrupt/aarch64/mod.rs
index d1263110..b59fda62 100644
--- a/qkernel/src/interrupt/aarch64/mod.rs
+++ b/qkernel/src/interrupt/aarch64/mod.rs
@@ -212,6 +212,16 @@ pub extern "C" fn exception_handler_el1h_sync(ptregs_addr:usize){
             let far = GetFarEL1();
             MemAbortKernel(ptregs_addr, esr, far, true);
         },
+        EsrDefs::EC_BRK => {
+            // ignore BRK for now.
+            if ptregs_addr == 0 {
+                panic!("exception frame is null pointer\n")
+            }
+            let ctx_p = ptregs_addr as *mut PtRegs;
+            let ctx_p = ctx_p.cast::<PtRegs>();
+            let ptregs = unsafe { &mut *ctx_p };
+            ptregs.pc += 4;
+        },
         _ => {
             panic!("unhandled sync exception from el1: {}\n", ec);
         }
```

**Comment by chl337** [comment]
It has to panic, as the syscall is not implemented. (I will get back to it shortly...)

**Comment by shrik3** [comment]
On Wed Mar 20, 2024 at 2:32 PM CET, chl337 wrote:
> It has to panic, as the syscall is not implemented.

it has nothing to do with skipping BRK exception (if I read your reply correctly)
...

And I don't think it's sys_io_setup. SysIoSetup is implemented and the config
EnableAIO is set to true. The error message is wrong I'm afraid (could be
another issue.).

The call_no is 439, the call is `sys_faccessat2`

https://elixir.bootlin.com/linux/v6.7/source/fs/open.c#L534

**Comment by shrik3** [comment]
> On Wed Mar 20, 2024 at 2:32 PM CET, chl337 wrote:
>  It has to panic, as the syscall is not implemented.
>
> The call_no is 439, the call is `sys_faccessat2`
>
> https://elixir.bootlin.com/linux/v6.7/source/fs/open.c#L534

This syscall is only added to linux kernel since 5.8. So it's new, and the
userspace should NOT expect it to be supported. In this case we should return
ENOSYS, instead of panic. The userspace library should be able to get around it.

Also shoudn't not care about users at all (and we can't)...

I suggest doing the same as  
https://github.com/QuarkContainer/Quark/pull/1080

**Comment by shrik3** [comment]
Let's skip faccessat (same as qkernel x86) With BRK exception skipped, the
program proceeds, until the VMM crashes with a substraction overflow...

```
---
diff --git a/qkernel/src/syscalls/syscalls.rs b/qkernel/src/syscalls/syscalls.rs
index 693cf0e3..9ecf879a 100644
--- a/qkernel/src/syscalls/syscalls.rs
+++ b/qkernel/src/syscalls/syscalls.rs
@@ -1037,7 +1037,7 @@ pub const SYS_CALL_TABLE: &'static [SyscallFn] = &[
     NotImplementSyscall,  //   436
     NotImplementSyscall,  //   437
     NotImplementSyscall,  //   438
-    NotImplementSyscall,  //   439
+    SysNoSys,             //   439 sys_faccessat2
     NotImplementSyscall,  //   440
     NotImplementSyscall,  //   441
     NotImplementSyscall,  //   442

```

**Comment by chl337** [comment]
> This syscall is only added to linux kernel since 5.8. So it's new, and the userspace should NOT expect it to be supported. In this case we should return ENOSYS, instead of panic.

I don't think returning a value that the actual syscall [would not return](https://elixir.bootlin.com/linux/latest/source/fs/open.c#L462) is a good idea. I thought about calling 
[SysFaccessat](https://github.com/QuarkContainer/Quark/blob/c62730bff6a2431960fa132304a991964620a9d3/qkernel/src/syscalls/sys_file.rs#L642) instead(which in itself is still an ugly thing to do).

## Result
![image](https://github.com/QuarkContainer/Quark/assets/119595393/ce026699-7cab-47ce-adf0-cd4808ecaf4f)

*More stuff in between*
_...and dead_
![image](https://github.com/QuarkContainer/Quark/assets/119595393/7efcbc5d-80a9-4fc8-9393-c2c632dd157f)

I don't have any idea right know, will return to it tomorrow.

**Comment by shrik3** [comment]
> I don't think returning a value that the actual syscall [would not return]
> (https://elixir.bootlin.com/linux/latest/source/fs/open.c#L462) is a good
> idea. [...]


You are mixing syscall return value and errno. Perhaps I didn't make it clear.

Take this for example

```
#include <unistd.h>
#include <stdio.h>
#include <sys/syscall.h>
#include <errno.h>
#include <string.h>

int main(void) {
	// make sure it works as expected:
	int ret;
	ret = syscall(SYS_write, 1, "hello world", 14);
	printf("\nSYS_write (%d) reutrns %d, errno = %d:%s\n", SYS_write, ret, errno, strerror(errno));
	// calling a non-exist (call it not implemented) syscall:
	ret = syscall(600, 1, "hello world", 14);
	printf("\nSYS_? (%d) reutrns %d, errno = %d:%s\n", 600, ret, errno, strerror(errno));
	return 0;
}
```

the output is:

```
hello world
SYS_write (1) reutrns 14, errno = 0:Success

SYS_? (600) reutrns -1, errno = 38:Function not implemented
```

The return value only indicates the "existence of error" and the errno
indicates the actual error. See man  syscall(2)

**Comment by chl337** [comment]
> I don't think returning a value that the actual syscall [would not return] (https://elixir.bootlin.com/linux/latest/source/fs/open.c#L462) is a good idea. [...]
> You are mixing syscall return value and errno.
Then I did misunderstood your suggestion.

**Comment by chl337** [comment]
I am afraid we are still missing something in _clone_. The point is that the error gets triggered when we try to execute a non-build-in shell command.
`docker run --rm -it --runtime=quark_d ubuntu sh` will work with the types of _echo_, _cd_, etc. but  crash with _ls_.

**Comment by shrik3** [comment]
On Thu Mar 21, 2024 at 5:58 PM CET, chl337 wrote:
 `docker run --rm -it --runtime=quark_d ubuntu sh` will work with the types of
> _echo_, _cd_, etc. but  crash with _ls_.

This isn't new XD. Do you have a debugging log on the crash?

**Comment by chl337** [comment]
Current log: 
[quark.log](https://github.com/QuarkContainer/Quark/files/14721208/quark.log)

**TL;DR** The parent process is killed because it tries to write to an address outside its mappings, and as a consequence, the child is also killed.

Maybe Interesting Points
---
* Parent
[INFO] [1/43c0140000|0] (1/1)------get call id sys_clone arg0:4111
[INFO] [1/43c0140000|0] Clone opts is CloneOptions { sharingOption: SharingOptions { NewAddressSpace: false, NewSignalHandlers: true, NewThreadGroup: true, TerminationSignal: Signal(11), NewPIDNamespace: false, NewUserNamespace: false, NewNetworkNamespace: false, NewFiles: true, NewFSContext: true, NewUTSNamespace: false, NewIPCNamespace: false }, Stack: 7f0bae7367d0, SetTLS: false, TLS: 0, ChildClearTID: false, ChildSetTID: false, ChildTID: fffffffe7fffffff, ParentSetTID: false, ParentTID: 0, Vfork: true, Untraced: false, InheritTracer: false }
[DEBUG] [1/43c01c0000|0] entering child task: kernelSp/PtRegs @ 43c01dfed0
    ................
* Child
[INFO] [1/43c01c0000|0] in the execve: the cmd is ls  
 envs is "HOSTNAME=d590b59e4345 TERM=xterm PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PWD=/ "
[DEBUG] [1/43c01c0000|0] VM: Map UPage VA-0x3ffffff000, PH-0x3ffffff000, flags-10000000000000000000000000000000000000000000010000010011
[DEBUG] [1/43c01c0000|0] VM: Map UPage VA-0x3fffffe000, PH-0x3fffffe000, flags-1100000000000000000000000000000000000000000011100001111
    ................
* Parent
[INFO] [1/43c0140000|0] (1/1)------Return[0] res is 2: call id sys_clone 
    ................
[DEBUG] [1/43c0140000|0] get data abort fault from el0
[INFO] [1/43c0140000|0] DFSC/IFSC == 0x5, FAR == 0xFFFFFFFF, acces-type fault == -w-, during address translation == false.
[ERROR] [1/43c0140000|0] VM: The map is 3fffffe000-8000000000 rw-p 00000000 00:00 0 
a000000000-a000001000 r--p 40004b2000 00:00 0                            [vvar]
a000001000-a000003000 r-xs 40004b3000 00:00 0                            [vdso]
df9cdee5000-df9cdf00000 r-xp 00000000 24:62b9a 38153114 
df9cdf00000-df9cdf01000 rwxp 0001b000 24:62b9a 38153114 
df9cdf10000-df9cdf12000 r--p 0001b000 24:62b9a 38153114 
df9cdf12000-df9cdf13000 rwxp 0001d000 24:62b9a 38153114 
df9cdf13000-df9cdf15000 rw-p 00000000 00:00 0                            [elf:static]
df9cdf15000-df9cdf36000 rw-p 00000000 00:00 0 
2b711fa96000-2b711fac0000 r-xp 00000000 24:62daa 38153642 
2b711fac0000-2b711fac1000 rwxp 0002a000 24:62daa 38153642 
2b711fac1000-2b711fac3000 rw-p 00000000 00:00 0 
2b711fad0000-2b711fad2000 r--p 0002a000 24:62daa 38153642 
2b711fad2000-2b711fad3000 rw-p 0002c000 24:62daa 38153642 
2b711fad3000-2b711fad4000 rwxp 0002d000 24:62daa 38153642 
2b711fae0000-2b711fc69000 r-xp 00000000 24:62dbc 38153660 
2b711fc69000-2b711fc78000 ---p 00189000 24:62dbc 38153660 
2b711fc78000-2b711fc7c000 r--p 00188000 24:62dbc 38153660 
2b711fc7c000-2b711fc7e000 rw-p 0018c000 24:62dbc 38153660 
2b711fc7e000-2b711fc8a000 rw-p 00000000 00:00 0 
7f0badf37000-7f0bae737000 rw-p 00000000 00:00 0                          [stack]
, fault address is not part of it.
[DEBUG] [1/43c0140000|0] VM: Entered HandleFault.
[INFO] [1/43c0140000|0] sendsignalLocked, signal:SignalInfo { Signo: 11, Errno: 0, Code: 2 }, group:false
    ................
* Child
[INFO] [1/43c01c0000|0] RunInterrupt[43c01c0000] ...
[INFO] [1/43c01c0000|0] Signal 9: terminating thread group
[INFO] [1/43c01c0000|0] RunExit[43c01c0000] ...

**Comment by shrik3** [comment]
> FAR == 0xFFFFFFFF

This looks dangerously like a dereference of an 32bit integer -1 (which is 0xffffffff). 

btw I think the problem is with sys_execv, not sys_clone. I could help later today.

**Comment by shrik3** [comment]
> FYI I tested the example in the _execve(2)_, and it passed.

Your mean, a bare execve test passed (not running from ubuntu sh) right?

**Comment by chl337** [comment]
> > FYI I tested the example in the _execve(2)_, and it passed.
> 
> Your mean, a bare execve test passed (not running from ubuntu sh) right?

A bare one.

**Comment by shrik3** [comment]
On Fri Mar 22, 2024 at 6:10 PM CET, chl337 wrote:
> A bare one.

I tested running a simple hello-world binary from `ubuntu sh`. And that failed
in the same way (FAR=0xffffffff).

**Comment by chl337** [comment]
I will close this as it is related to #1210.
